many: `systemd-firstboot` support

supakeen opened the pull request:
This PR introduces a bunch of preliminary work to get `systemd-firstboot` to work with the Fedora Minimal spin though it is set up that it could be configured with other image types as well.

For the eventual interface I *think* we don't want each image type to define all the parts separately but to instead have a single flag in the image config to 'SystemdFirstBootable' or such.

The work is not yet done but this PR serves for documentation of encountered problems and such.

---

## Problems, challenges, fun, quirks

1. `systemd-firstboot --image=...` works great, everything is configurable.

However when booting an image more software runs and things start to get funky.

1. `systemd-tmpfiles` is configured on Fedora to create (amongst others) the `/etc/locale.conf`, `/etc/vconsole.conf` files (see: `/usr/lib/tmpfiles.d/etc.conf`). `systemd-tmpfiles` is explicitly requested to run *before* `systemd-firstboot`. This needs to be changed either in the Fedora-shipped RPMs or it needs to be worked around at build time (patch the `tmpfiles.d` configurations? change the order?) otherwise those options cannot be configured.
2. The `systemd-firstboot` prompt is 'broken'. To start with we get a bunch of output that goes straight through the prompt, perhaps some ordering issue; for now a `PreExecStart=/bin/sleep 1` fixes this but is not the right thing to do. 
3. The `systemd-firstboot` prompt is 'broken'. Initial keypresses do not work on the console. Potentially we'd want to wait for `udev-settle` or not at all.
4. The `root` user needs to be handled, currently it is created by either `systemd-tmpfiles` or `systemd-sysusers` (haven't looked yet) *before* `systemd-firstboot` runs since.

supakeen committed:
fedora: unset hostname for fedora minimal

Unset the hostname so it isn't written to disk by default, allowing
`systemd-firstboot` to do so.

Signed-off-by: Simon de Vlieger <supakeen@redhat.com>

supakeen committed:
fedora: unset locale for fedora minimal

Unset the locale so it isn't written to disk by default, allowing
`systemd-firstboot` to do so.

Signed-off-by: Simon de Vlieger <supakeen@redhat.com>

supakeen committed:
fedora: unset timezone for fedora minimal

Unset the timezone so it isn't written to disk by default, allowing
`systemd-firstboot` to do so.

Signed-off-by: Simon de Vlieger <supakeen@redhat.com>

supakeen committed:
fedora: disable initial-setup for minimal

Disable initial setup as `systemd-firstboot` can now configure the image
instead.

Signed-off-by: Simon de Vlieger <supakeen@redhat.com>

achilleas-k commented on pkg/manifest/os_test.go:
```diff
@@ -248,3 +248,21 @@ func TestFirstBootDoesNotIncludeMachineIdStage(t *testing.T) {
 	require.Nil(t, st)
 }
+
+func TestLanguageIncludesLocaleStage(t *testing.T) {
+	os := NewTestOS()
+
+	os.Language = "en_US.UTF-8"
+
+	pipeline := os.serialize()
+	st := findStage("org.osbuild.locale", pipeline.Stages)
+	require.NotNil(t, st)
+}
+
+func TestLanguageDoesNotIncludeLocaleStage(t *testing.T) {
+	os := NewTestOS()
+
+	pipeline := os.serialize()
+	st := findStage("org.osbuild.locale", pipeline.Stages)
+	require.Nil(t, st)
+}
```
This commit can be squashed with the first one.  They even have almost the same message already.

supakeen replied:
Done.

achilleas-k commented on pkg/distro/fedora/distro.go:
```diff
@@ -442,6 +442,8 @@ var (
 				// Overwrite the default Grub2 timeout value.
 				Timeout: 5,
 			},
+			// Unset hostname so it isn't written (necessary for systemd-firstboot).
+			Hostname: common.ToPtr(""),
```
Wouldn't it be more correct to set it to `nil` (or keep it unset)?

supakeen replied:
Setting it to `nil` does the wrong thing when the merge between image configs gets done; it then gets turned into the distro default.

achilleas-k commented on pkg/distro/fedora/distro.go:
```diff
@@ -444,6 +444,8 @@ var (
 			},
 			// Unset hostname so it isn't written (necessary for systemd-firstboot).
 			Hostname: common.ToPtr(""),
+			// Unset locale so it isn't written (necessary for systemd-firstboot).
+			Locale: common.ToPtr(""),
```
Same as previous comment about hostname.

Setting this to `ptr("")` will make the stage run and write `""`, no?

supakeen replied:
Setting it to `nil` does the wrong thing when the merge between image configs gets done; it then gets turned into the distro default.

The `os` pipeline verifies that the value is non-empty (e.g. `if p.Language != ""`) as a conditional for writing the stage so it doesn't add the stage at all.

achilleas-k reviewed (commented):
This is really nice.  Good cleanup and unification of option handling.

Couple of comments below.

Any reason it's kept draft?  I think we can merge it and continue working on the firstboot process from here.

supakeen commented:
Rebased, addressed @achilleas-k's comments.



> Any reason it's kept draft? I think we can merge it and continue working on the firstboot process from here.

Everything before the *last* commit is mergeable. I can create another PR with all the commits minus the last one because I think those are fine. Would that be something you want?

The last commit makes people unable to actually configure their system on *boot* for the reasons listed in the topic post; you can't configure a root password and systemd-tmpfiles pre-writes these files. I haven't had time to jerryrig that temporarily and anything we do for that will need to go into Fedora :)

supakeen committed:
fedora: remove plymouth for fedora minimal

With plymouth enabled the terminal behaves very weirdly during
systemd-firstboot. Keypresses are not registered.

More investigation is necessary to figure out *why* this happens but for
now this is a good workaround.

We still have some output crossing, so that's next.

Signed-off-by: Simon de Vlieger <supakeen@redhat.com>

supakeen commented:
@achilleas-k Thanks for the approval: this needs to be scoped so it only applies to Fedora 43 (rawhide) because it relies on some things that exist there, unless I can confirm it to work on Fedora 42. I hope to get to those tidbits tomorrow.

After that it's going to need approval from more people than just code owners, I think, mostly Peter, Paul, and Adam (not tagging on purpose). The first two due to it being a significant change in the minimal fedora artifact. The last one due to having to make sure that the QA systems can verify this, since minimal is release blocking in fedora.

supakeen commented:
Still active, I just need to find time to remap/version gate this still.

AdamWill commented:
Is this no longer planned? We have an [os-autoinst-distri-fedora ticket](https://forge.fedoraproject.org/quality/os-autoinst-distri-fedora/issues/367) that depends on it, so I'm checking in.

supakeen commented:
> Is this no longer planned? We have an [os-autoinst-distri-fedora ticket](https://forge.fedoraproject.org/quality/os-autoinst-distri-fedora/issues/367) that depends on it, so I'm checking in.

We can land this without this PR but I need to discuss with Peter when we want to enable this.
